src/backend/distributed/planner/multi_planner_utils.c

```diff
@@ -0,0 +1,863 @@
+ * joined on their partition keys.
+ *
+ * The function returns true if all relations are joined on their partition keys.
+ * Otherwise, the function returns false. Since reference tables do not have partition
+ * keys, we skip processing them. Also, if the query includes only a single non-reference
+ * distributed relation, the function returns true since it doesn't make sense to check
+ * for partition key equality in that case.
+ *
+ * In order to do that, we invented a new equivalence class namely:
+ * AttributeEquivalenceClass. In very simple words, a AttributeEquivalenceClass is
+ * identified by an unique id and consists of a list of AttributeEquivalenceMembers.
+ *
+ * Each AttributeEquivalenceMember is designed to identify attributes uniquely within the
+ * whole query. The necessity of this arise since varno attributes are defined within
+ * a single level of a query. Instead, here we want to identify each RTE_RELATION uniquely
+ * and try to find equality among each RTE_RELATION's partition key.
+ *
+ * Each equality among RTE_RELATION is saved using an AttributeEquivalenceClass where
+ * each member attribute is identified by a AttributeEquivalenceMember. In the final
+ * step, we try generate a common attribute equivalence class that holds as much as
+ * AttributeEquivalenceMembers whose attributes are a partition keys.
+ *
+ * AllRelationsJoinedOnPartitionKey uses both relation restrictions and join restrictions
+ * to find as much as information that Postgres planner provides to extensions. For the
+ * details of the usage, please see GenerateAttributeEquivalencesForRelationRestrictions()
+ * and GenerateAttributeEquivalencesForJoinRestrictions()
+ */
+bool
+AllRelationsJoinedOnPartitionKey(PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RelationRestrictionContext *restrictionContext =
+		plannerRestrictionContext->relationRestrictionContext;
+	JoinRestrictionContext *joinRestrictionContext =
+		plannerRestrictionContext->joinRestrictionContext;
+
+	List *relationRestrictionAttributeEquivalenceList = NIL;
+	List *joinRestrictionAttributeEquivalenceList = NIL;
+	List *allAttributeEquivalenceList = NIL;
+	AttributeEquivalenceClass *commonEquivalenceClass = NULL;
+	uint32 referenceRelationCount = ReferenceRelationCount(restrictionContext);
+	uint32 totalRelationCount = list_length(restrictionContext->relationRestrictionList);
+	uint32 nonReferenceRelationCount = totalRelationCount - referenceRelationCount;
+	ListCell *commonEqClassCell = NULL;
+	ListCell *relationRestrictionCell = NULL;
+	Relids commonRteIdentities = NULL;
+
+	/*
+	 * If the query includes a single relation which is not a reference table,
+	 * we should not check the partition column equality.
+	 * Consider two example cases:
+	 *   (i)   The query includes only a single colocated relation
+	 *   (ii)  A colocated relation is joined with a reference
+	 *         table where colocated relation is not joined on the partition key
```

Comment:
But what if there's more than one non-reference table involved?  We need to be safe against this anyway.